sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsShuffleHeartbeatManager.scala

```diff
@@ -79,31 +115,71 @@ class RapidsShuffleHeartbeatManager extends Logging {
    * @return `RapidsExecutorUpdateMsg` with new executors, since the last heartbeat was received.
    */
   def executorHeartbeat(id: BlockManagerId): RapidsExecutorUpdateMsg = synchronized {
-    val lastRegistration = lastRegistrationSeen.get(id)
+    val lastRegistration = executorRegistrations.get(id)
     if (lastRegistration == null) {
       throw new IllegalStateException(s"Heartbeat from unknown executor $id")
     }
 
+    removeDeadExecutors(getCurrentTimeMillis, Option(id))
+
     val newExecutors = new ArrayBuffer[BlockManagerId]
-    val iter = executors.zipWithIndex.reverseIterator
+    val iter = executors.reverseIterator
     var done = false
     while (iter.hasNext && !done) {
-      val (entry, index) = iter.next()
+      val entry = iter.next()
 
-      if (index > lastRegistration.getValue) {
+      if (entry.registrationOrder >= lastRegistration.lastRegistrationOrderSeen.getValue) {
         if (entry.id != id) {
+          logDebug(s"Found new executor (to $id): $entry while handling a heartbeat.")
           newExecutors += entry.id
         }
       } else {
-        // We are iterating backwards and have found the last index previously inspected
-        // for this peer. We can stop since all peers below this have been sent to this peer.
+        // We are iterating backwards and have found the last previously inspected executor
+        // since `registrationOrder` monotonically increases and follows insertion order.
+        // We can stop since all peers below this have been sent to the heartbeating peer.
         done = true
       }
     }
 
-    lastRegistration.setValue(executors.size - 1)
+    // update this executor's registration with a new heartbeat time, and that last order
+    // from the executors list, indicating the order we should stop at next time
+    lastRegistration.lastHeartbeatMillis.setValue(getCurrentTimeMillis)
+    lastRegistration.lastRegistrationOrderSeen.setValue(registrationOrder)
+
+    // since we updated our heartbeat, update our min-heap
+    leastRecentHeartbeat.priorityUpdated(lastRegistration.id)
     RapidsExecutorUpdateMsg(newExecutors.toArray)
   }
+
+  // the timeout amount has elapsed, assume this executor is dead
+  private def isStaleHeartbeat(hbMillis: Long, currentTime: Long): Boolean = {
+    (currentTime - hbMillis) > heartbeatTimeoutMillis
+  }
+
+  // must be called holding the object lock
+  private def removeDeadExecutors(currentTime: Long,
+                                  heartbeatingExec: Option[BlockManagerId]): Unit = {
```

Comment:
I'm confused about this parameter.  Why are we special-casing the scenario where the current heartbeat is for an executor that is stale?  We need to gracefully handle the case where an executor heartbeats and is already considered stale, so I don't see the need to special case this.

Reply:
I did that because if the current executor is telling me it is not stale, I thought it was wasteful to remove and re-add. Yes the graceful rejoin prompted me to do that. I am planning on handling that part as suggested. The other comment made me think, if we have several executors all GCing and missing heartbeats, then the re-add is going to happen anyway for most of those.. so this isn't probably saving is much.

Reply:
Exactly.  If there's a lot of executors then the chance that _this exact executor heartbeat_ is when an executor would have been declared stale is exceedingly small.  Since we have to gracefully handle the case where the executor is already considered stale when it heartbeats, I personally would simply make trimming stale executors the first thing any registration or heartbeat handler code does before it proceeds.